lib/logstash/inputs/beats.rb

```diff
@@ -207,9 +210,229 @@ def invoke(connection, &block)
+                                                  map,
+                                                  identity_stream,
+                                                  @codec_transformer,
+                                                  @queue))
+      end
+    end
+
+    # OOB call to flush the codec buffer, 
+    #
+    # This method is a bit tricky to decide when to be called, in the current case,
+    # this will be call on any exception raised, either is a circuit breaker or the
+    # remote host closed the connection, its better to make sure we clear their
+    # data and create duplicates then losing the data.
+    def flush(&block)
+      @logger.debug? && @logger.debug("Beats input, out of band call for flushing the content of this connection",
+                                      :peer => @connection.peer)
+
+      @codec.flush(&block)
+    end
+  end
+
+  # Base Transform class, expose the plugin decorate method_defined?
+  # apply the tags and make sure we copy the beat hostname into `host`
+  # for backward compatibility.
+  class EventTransformCommon
+    def initialize(input)
+      @input = input
+      @logger = input.logger
+    end
+
+    # Copies the beat.hostname field into the host field unless
+    # the host field is already defined
+    def copy_beat_hostname(event)
+      host = event["beat"] ? event["beat"]["hostname"] : nil
+      if host && event["host"].nil?
+        event["host"] = host
+      end
+    end
+
+    # This break the `#decorate` method visibility of the plugin base
+    # class, the method is private and we cannot access it, but well ruby
+    # can let you do all the wrong thing.
+    #
+    # I think the correct behavior would be to allow the plugin to return a 
+    # `Decorator` object that we can pass to other objects, since only the 
+    # plugin know the data used to decorate. This would allow a more component
+    # based workflow.
+    def decorate(event)
+      @input.send(:decorate, event)
+    end
+
+    def transform(event)
+      copy_beat_hostname(event)
+      decorate(event)
+      event
+    end
+  end
+
+  # Take the the raw output from the library, decorate it with
+  # the configured tags in the plugins and normalize the hostname
+  # for backward compatibility
+  #
+  #
+  # @see [Lumberjack::Beats::Parser]
+  #
+  class JsonTransform < EventTransformCommon
+    def transform(event)
+      super(event)
+      event.tag("beats_input_no_codec_applied")
+      event
+    end
+  end
+
+  # Take the extracted content from the codec, merged with the other data coming
+  # from beats, apply the configured tags, normalize the host and try to coerce
+  # the timestamp if it was provided in the map.
+  class CodecTransform < EventTransformCommon
+    def transform(event, map)
+      ts = coerce_ts(map.delete("@timestamp"))
+
+      event["@timestamp"] = ts unless ts.nil?
+      map.each { |k, v| event[k] = v }
+      super(event)
+      event.tag("beats_input_codec_applied")
```

Comment:
I think its better in that case to add it all the time, since it will give more visibility into what the plugins did.

I wanted to expose if the beats input went through a json map or a codec decode since its not obvious all the time, but I am + 1 to add the name of the codec used in the tag.